src/http.ts

```diff
@@ -1,4 +1,12 @@
-import fetch from "node-fetch";
+const fetch: typeof globalThis.fetch = async (...args) => {
+  if (globalThis?.fetch) {
+    return await globalThis.fetch(...args);
+  } else {
+    const nodeFetch = (await dynamicImport("node-fetch")).default;
+    return await nodeFetch(...args);
+  }
+};
```

Comment:
> ...downstream lemmy-ui would also probably have to do this hack just to support node-fetch v3

Fortunately, downstream users won't have to use the hack. I tested the lemmy-ui with the updated client. With the exception of a bit of fiddling with yalc and docker to actually be able to use the locally modified client, no extra code was needed. I tested the UI on chrome, firefox, and edge and it worked in each.

I also made a small program to run on node that depends on the client to test it. The dynamic import hack wasn't needed there either. Just to be thorough, I tested the client on a bunch of versions of node. For each major LTS version 12 and up, it works on:

v12: 12.20.2 and higher
v14: 14.14.0 and higher
v16: every version I tested
v18: every version I tested

That should hopefully address what seems to be your main concern with the approach I took. As to why I decided to do this import weirdness to use v3 instead of just sticking v2, to make a long story short, I couldn't get it to work any other way, and not for lack of trying. I'm aware this sounds like a cop out, so I'll explain and hopefully you'll see the method in my madness.

The image upload needs to use form data to work. There is a form data API that's supported most modern browsers. However, there's nothing like that as part of node, not even as an experimental feature. I'm assuming the reason you're using `node-fetch` and not the globally defined fetch API is because you want your client to work on both browsers and node. After a bit of digging, I found that [`form-data`](https://www.npmjs.com/package/form-data) was suggested for that purpose. It hasn't received an update in 2 years, so we're already off to a good start. I tried passing buffers and blobs to the form data append, and every time I tried to make a request I got an error about the request being invalid. I had to try multiple different blob APIs, because there isn't a built in global Blob API in node in versions earlier than 18. I tried [`fetch-blob`](https://www.npmjs.com/package/fetch-blob) as well as the [Blob from node's `buffer` module](https://nodejs.org/api/buffer.html#class-blob). Neither of them worked. With `fetch-blob`, I even had to use an out of date version because everything above version 3 has the same ESM issue I was trying to avoid, but unlike `node-fetch`, where version 2 still receives updates, the last compatible version is 2 years old. I couldn't even try to new up [File](https://developer.mozilla.org/en-US/docs/Web/API/File) since there doesn't seem to be a npm module for a clone of that API like there is for form data and blobs; without any dependencies, the best node's got so far is an experimental feature in the `buffer` module that's only available on versions 18.13 and up.

I found another form data module to try, [`formdata-node`](https://www.npmjs.com/package/formdata-node). Its last release was 7 months ago and it's github repo received updates within the past few days, which seemed more promising than the last module. It also has its own implementation of Blob and File that are close to the browser types. However, this one requires adding yet another dependency, [`form-data-encoder`](https://www.npmjs.com/package/form-data-encoder), in order for it to work with `node-fetch` v2. Once I added the dependencies and tried running the node program I made to test it, I discovered that these modules have the ESM issue that `node-fetch` v3 has, which seems to defeat the point. Even if this isn't an issue on a lower (and no longer supported) major version, using the encoder requires using node's `stream` module, so some dynamic importing weirdness would still be required to get it to work on both browsers and node.

`node-fetch` v3 includes implementations of `File`, `Blob`, and `FormData`; this eliminates the need to juggle versions of different npm packages in order to maintain compatibility. Once I figured out how to prevent typescript from compiling dynamic imports into requires, getting the function working was pretty straightforward. This was after several hours of banging my head against the wall trying to get this working for v2.

Between my efforts to implement the feature using v2 bearing no fruit, the fact that dependents of the library don't have to do the dynamic import hack to get it working, and the fact that it works on every major browser I could test on as well as many different LTS version of node (including all of the versions of LTS since v16) means I think using v3 is appropriate.

Reply:
> I'm assuming the reason you're using node-fetch and not the globally defined fetch API is because you want your client to work on both browsers and node

Yup that's correct, the same fetch is used for both the back and front ends.

I'm sorry to have to add more work for you, but I **really** don't wanna have to deal with this node-v3 import nightmare, for what should be a simple http library. This type of custom importing code supporting specific browsers and node versions is going to be that much more pointless work for all of us to maintain.

Could you try a library like cross-fetch and see if it works? https://github.com/lquixada/cross-fetch  . There's probably also a few other isomorphic fetch libraries that we could try before giving into this.

Reply:
I got it working with `cross-fetch` and `form-data`.